### PR TITLE
fix: normalize all whitespace types in SQL query cleaning

### PR DESCRIPTION
**Pull request by rafaelpeter03**
# Fix: Normalize all whitespace types in SQL query cleaning

## Problem
The current SQL cleaning logic in `GoogleBigQueryTools._run_sql()` causes syntax errors when queries contain mixed whitespace (tabs, multiple spaces, line breaks), resulting in concatenated keywords without proper spacing.

**Example Error:**
```sql
-- Input query
SELECT 
    SUM(revenue) AS total_revenue
FROM 
    sales

-- After cleaning (broken)
SELECT     SUM(revenue) AS total_revenueFROM     sales
--                                        ↑ Missing space!

-- BigQuery error
Syntax error: Expected end of input but got identifier "sales" at [1:52]
```

## Solution
Replace character-by-character replacements with regex-based whitespace normalization:

```python
# Added import
from re import sub

# Old approach
# cleaned_query = sql.replace("\\n", " ").replace("\n", "").replace("\\", "")

# New approach - normalizes ALL whitespace to single space
cleaned_query = sub(
    r"\s+",
    " ",
    sql.replace("\\n", "\n")
       .replace("\\r", "\r")
       .replace("\\t", "\t")
       .replace("\\", ""),
).strip()
```

## Benefits
- ✅ Handles all whitespace types (spaces, tabs, newlines, carriage returns)
- ✅ Prevents keyword concatenation errors
- ✅ More robust for LLM-generated queries
- ✅ Fully backward compatible

## Testing
```python
# Multi-line with tabs
"SELECT\t\tCOUNT(*)\nFROM\t\tcustomers"
# → "SELECT COUNT(*) FROM customers"

# Mixed whitespace
"SELECT   \n\n   product_name  \t FROM   \r\n  products"
# → "SELECT product_name FROM products"
```

All queries now execute successfully without syntax errors.

### CODE REVIEW

File: libs/agno/agno/tools/google_bigquery.py [hunk 1]
```diff
@@ -106,7 +107,14 @@ def _run_sql(self, sql: str) -> str:
...
         try:
             log_debug(f"Running Google SQL |\n{sql}")
-            cleaned_query = sql.replace("\\n", " ").replace("\n", "").replace("\\", "")
+            cleaned_query = sub(
+                r"\s+",
+                " ",
+                sql.replace("\\n", "\n")
+                .replace("\\r", "\r")
+                .replace("\\t", "\t")
+                .replace("\\", ""),
```
Comment (manuhortet) [inline libs/agno/agno/tools/google_bigquery.py:116]: I'm not completely sure about this final line. It will remove all backslashes, but some we may need to stay. I'm thinking about the ones in regex patterns or escaped quotes within SQL strings.

What do you think?

**Review by manuhortet — COMMENTED** [review]
Looks good @rafaelpeter03, thanks for the contribution! I have one comment, let's clarify that and we can merge this one.

File: libs/agno/agno/tools/google_bigquery.py (hunk 1 above)
Comment (rafaelpeter03) [reply libs/agno/agno/tools/google_bigquery.py:116]: You're right, I ended up having other limitations in using BigQuery's real capacity using agno's native tool, I developed a fork of it with other implementations and I'm validating its operation.

The native lib has 2 very considerable errors, which is that it breaks the query by removing spaces that shouldn't be removed and another one which is that when the agent makes a mistake in the query it doesn't give him back what he got wrong, making him experience a lot of trial and error.

Do you think it's better that I use this branch with the suggestion to fix of these two flaws or that we close this one and I focus on implementing it with more capabilities and features?

Comment (manuhortet) [reply libs/agno/agno/tools/google_bigquery.py:116]: > when the agent makes a mistake in the query it doesn't give him back what he got wrong, making him experience a lot of trial and error.

That error feedback would be a great addition.

You can decide what to do, both ways sound good. If you are already working on the re-implementation, I'll be happy to review and get that merged too if you open a PR with it! but your call.

Thanks for the work here!

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR fixes SQL query cleaning in Google BigQuery integration by replacing the previous string-based whitespace normalization with a comprehensive regex-based approach. The new implementation handles all whitespace types (spaces, newlines, carriage returns, tabs) including their escaped variants, providing more robust query sanitization.

## Architecture

The PR improves SQL query processing through better whitespace handling:

- **Regex-Based Normalization**: Replaces multiple `.replace()` calls with single `sub(r"\s+", " ", ...)` regex pattern
- **Comprehensive Whitespace Handling**: Handles all whitespace types (\n, \r, \t) plus their escaped variants (\\n, \\r, \\t)
- **Output Cleaning**: Adds `.strip()` to remove leading/trailing whitespace from final query
- **Import Addition**: Adds `from re import sub` for regex substitution

The implementation addresses a limitation in the original `_run_sql()` method where various whitespace characters in SQL queries could cause issues with BigQuery execution. The fix ensures consistent query formatting regardless of the source of the whitespace.

## Issues

### Critical

No critical issues found. The implementation is correct and improves the existing functionality.

### Warnings

**libs/agno/agno/tools/google_bigquery.py:110-119** — **regex pattern could be more explicit** — The regex pattern might not handle all edge cases:

```python
cleaned_query = sub(
    r"\s+",  # Whitespace characters
    " ",
    sql.replace("\\n", "\n")  # Handle escaped newlines first
    .replace("\\r", "\r")
    .replace("\\t", "\t")
    .replace("\\", ""),
).strip()
```

**Suggested fix**: Consider using a more explicit pattern like `r"[\\s\\n\\r\\t]+"` or pre-processing with `codecs.decode()` if the SQL comes from certain sources.

**libs/agno/agno/tools/google_bigquery.py:107-119** — **order of replacements could matter** — The chained `.replace()` calls might not handle all cases optimally:

```python
# Current order:
# 1. Replace escaped sequences (\\n -> \n, etc.)
# 2. Apply regex to collapse all whitespace
# 3. Remove remaining backslashes
# 4. Strip final result

# But what if: "\\t\\n" becomes "\t \n" then "\t"?
```

**Suggested fix**: Test the replacement order with various whitespace combinations to ensure consistent results.

### Suggestions

**Missing test coverage** — The PR doesn't include tests for the improved whitespace handling:

```python
# No tests added to verify:
# - Multiple whitespace types are properly normalized
# - Escaped characters are handled correctly
# - Leading/trailing whitespace is removed
# - Complex whitespace patterns are cleaned
```

**Suggested fix**: Add unit tests with various SQL inputs containing different whitespace patterns.

**Performance consideration** — The new implementation adds regex processing overhead:

```python
# Old approach: Multiple .replace() calls (string operations, fast)
# New approach: Regex substitution (slower but more comprehensive)
# Trade-off: Accuracy vs. Performance
```

**Suggested fix**: Consider performance implications for large SQL queries or high-frequency operations. Document any performance characteristics.

**Limited scope improvement** — The fix only applies to Google BigQuery tool:

```python
# What about:
# - Other SQL tools (Snowflake, Redshift, etc.)?
# - General SQL cleaning utilities?
# - Consistent whitespace handling across the codebase?
```

**Suggested fix**: Consider creating a shared SQL cleaning utility or applying this fix to other SQL-based tools if similar issues exist.

**Error handling** — No additional error handling for regex operations:

```python
cleaned_query = sub(
    r"\s+",
    " ",
    sql.replace("\\n", "\n")  # What if replace fails?
    .replace("\\r", "\r")
    .replace("\\t", "\t")
    .replace("\\", ""),
).strip()
# No try-except, no validation of cleaned_query
```

**Suggested fix**: Add error handling for regex operations and validate the output before sending to BigQuery.

## Cross-file Impact

**Public API**: No API changes, only internal implementation improvement.

**Behavior Change**: Improves SQL query cleaning for more robust whitespace handling.

**Breaking Changes**: None. This is a bug fix.

**Test Coverage**: No new tests added.

**Dependencies**: Adds `re` module import (Python standard library).

## Assessment

✅ Approve

This PR is a focused, correct bug fix that improves SQL query cleaning in the Google BigQuery tool. The implementation replaces a limited string-based approach with a more comprehensive regex-based solution that handles all whitespace types including their escaped variants.

The fix is well-designed:
- **Comprehensive Coverage**: Handles spaces, newlines, carriage returns, tabs and their escaped forms
- **Clean Implementation**: Uses regex pattern matching which is more robust than multiple `.replace()` calls
- **Output Quality**: Adds `.strip()` to ensure clean final queries
- **Minimal Scope**: Only modifies the specific function that needs improvement

While minor improvements could be made (test coverage, performance consideration, broader application), these are not blocking concerns. The core functionality is sound and provides immediate value by fixing SQL query formatting issues that could cause BigQuery execution problems.

The regex approach (`r"\s+"`) is appropriate for normalizing whitespace and the sequential handling of escaped characters followed by regex substitution ensures proper conversion. This is a pragmatic bug fix that should be merged.

**Comment by harshsinha03** [comment]
Hey, appreciate the effort here! This was already fixed on main via #5957, so I'm closing this as resolved. Thank you for using Agno!
